### ISSUE TITLE
[Bug/Question]: VerneMQ consistently degraded performance after a few days

### ISSUE THREAD

**Issue by sovaa** [vernemq/vernemq#2085]
### Environment

- VerneMQ Version: 1.12.6.2
- OS:  Ubuntu 20.04.5 LTS
- Erlang/OTP version (if building from source): Erlang/OTP 23 [erts-11.2.2.4]
- Cluster size/standalone: 4 nodes (256 gb ram and 48/96 cores/threads each)

### Current Behavior

After a few days, metrics start to become less and less exported, until almost no metrics are exposed anymore. After a few days also there's issues with clients randomly not being able to subscribe, growing more and more common after a few days. Required to delete the data folder and recreate the cluster to fix the issue.

### Expected behaviour

Cluster should run stable even after many days.

### Configuration, logs, error output, etc.

We're publishing about 3 million messages per day to around 100k different clients (client comes and goes, only about 10-15k are connected at the same time), on 4 different ws listeners. Each client subscribes to their own queue and we don't do any fanouts to groups of clients.

My only idea is it's maybe something to do with `upgrade_outgoing_qos`, or the message storage? We don't need to persist messages, so would disabling storage completely as mentioned in #1693 be okay?

Changes made to vernemq.conf:

```markdown
allow_anonymous = off
allow_register_during_netsplit = on
allow_publish_during_netsplit = off
allow_subscribe_during_netsplit = on
allow_unsubscribe_during_netsplit = on
coordinate_registrations = off
max_client_id_size = 100
persistent_client_expiration = 1d
max_inflight_messages = 0
max_online_messages = 50
max_offline_messages = 0
max_message_size = 536870
upgrade_outgoing_qos = on
listener.max_connections = 1000000
listener.nr_of_acceptors = 1000
listener.tcp.allowed_protocol_versions = 5
listener.ws.allowed_protocol_versions = 5
systree_enabled = off
plugins.vmq_diversity = on
metadata_plugin = vmq_swc
vmq_diversity.auth_redis.enabled = on
vmq_bcrypt.pool_size = auto
erlang.async_threads = 192
leveldb.maximum_memory.percent = 40
```

### Code of Conduct

- [X] I agree to follow the VerneMQ's Code of Conduct

**Comment by sovaa** [comment]
Each time the graphs look normal again is when the data folder have been removed and the cluster has been recreated:

![VerneMQ-Dashboard-Grafana](https://user-images.githubusercontent.com/213085/218290343-a5ac07c2-c94a-40e8-a57d-6df98d0482dd.png)

![VerneMQ-Dashboard-Grafana (1)](https://user-images.githubusercontent.com/213085/218290350-ae5be557-14b2-432c-a0d3-05206dd55d1b.png)

**Comment by ioolkos** [comment]
Firstly, as with any considerable use case, please consider supporting and sponsoring the VerneMQ project.

What to do

- I'd compile Verne with at least OTP 24.3
- are you sure clients do not request a persistent session?
- check the logs for dropped messages
- check the logs for sync issues
- check the logs
- give the online queue more elasticity (max_online_messages is 50, default is 1000)
- do not set `max_inflight` to 0 but to some specific value (200 or even 1000 if you have slow consumers)
- review your load balancer strategy. Is it round robin? again: are you sure you don't have persistent sessions for which a sticky session approach would be more apt?
- use various `recon` functions to trace for most working processes and process mailbox sizes (`proc_window` etc)
- why is `upgrade_outgoing_qos` enabled? does this mean you have persistent sessions at the outgoing end?
- did you try other settings than max_offline_messages = 0? did you see queues collect messages?
- do you use shared subscribers?

Broker behaviour is defined by input and configuration. Performance degradation is not a bug, unless your logs point to a specific issue. If it's a bug, it must be related to v5, since you only allow v5 connections.

TL;DR: logs.

Also note: you should be able to handle a scenario like this with much smaller sized instances. (that's just my free opinion added, not having further details of your use case)

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by sovaa** [comment]
Thanks a lot for the tips. Changes made (will see if it helps in a few days):

* changed `max_inflight_messages` to `500`, 
* disabled `upgrade_outgoing_qos` (only one not often triggered use case that couldn't send QoS 1, so this option was enabled, not important to be QoS 1 though),
* changed `max_online_messages` to `1000`.

We do have sticky sessions on the load balancer (`ip_hash`). We do not use shared subscribers (no `$share/sharename/topic` topics).

We do have persistent sessions if I understand correctly (`clean=false`, unique client IDs). Clients are usually only connected for about an hour a day. A client can connect multiple times and subscribe to the same topic, but all have a unique client ID.

**Comment by ioolkos** [comment]
For details about persistent sessions please refer to the MQTT specification.

Yes, every MQTT client has a unique ClientID  (otherwise they'd create races on the broker). The important point here is to not use a new ClientID for every new CONNECT of the same client. (adding a timestamp to the ClientID, or "ideas" like that).

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by ioolkos** [comment]
@sovaa any updates so far?

If you need more thorough commercial support, feel free to write to contact @ vernemq.com

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by sovaa** [comment]
We've now also changed to unique Client IDs (but no random suffix), and we skipped using persistent sessions. 

It seems okay last couple of days at least. Thanks a lot for the help!

**Comment by ioolkos** [comment]
Good to hear. Yes, random ClientIds create "forgotten" sessions/queues on the broker.
A reminder to all our users: Please consider supporting the VerneMQ project.

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.
